File: samples/dynamic-mcp-clients/dynamic-mcp-clients-assistant/src/main/java/io/quarkiverse/langchain4j/sample/assistant/service/McpToolProviderSupplier.java
package io.quarkiverse.langchain4j.sample.assistant.service;

import dev.langchain4j.mcp.McpToolProvider;
import dev.langchain4j.service.tool.ToolProvider;
import dev.langchain4j.service.tool.ToolProviderRequest;
import dev.langchain4j.service.tool.ToolProviderResult;
import jakarta.enterprise.context.ApplicationScoped;
import jakarta.enterprise.inject.Produces;

import java.util.ArrayList;

@ApplicationScoped
public class McpToolProviderSupplier implements ToolProvider {

    private final McpToolProvider mcpToolProvider;

    public McpToolProviderSupplier() {
        this.mcpToolProvider = McpToolProvider.builder().mcpClients(new ArrayList<>()).build();
    }

    @Override
    public ToolProviderResult provideTools(ToolProviderRequest request) {
        return mcpToolProvider.provideTools(request);
    }

    @Produces
    @ApplicationScoped
    public McpToolProvider mcpToolProvider() {
        return mcpToolProvider;
    }
}

File: samples/chatbot-easy-rag-kotlin/src/main/kotlin/io/quarkiverse/langchain4j/sample/chatbot/tools/McpToolProviderSupplier.kt
package io.quarkiverse.langchain4j.sample.chatbot.tools

import dev.langchain4j.mcp.McpToolProvider
import dev.langchain4j.mcp.client.McpClient
import dev.langchain4j.service.tool.ToolProvider
import io.quarkiverse.langchain4j.mcp.runtime.McpClientName
import jakarta.enterprise.context.ApplicationScoped
import jakarta.inject.Inject
import java.util.function.Supplier

@ApplicationScoped
class McpToolProviderSupplier : Supplier<ToolProvider> {

    @Inject
    @McpClientName("time")
    private lateinit var timeClient : McpClient

    override fun get(): ToolProvider {
        return McpToolProvider.builder()
            .mcpClients(timeClient)
            .build()
    }

}
